### PR TITLE
Construct the hook kwargs inside each loop

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Resolve TODO. Follow-up to #11466.

This encapsulates logic within each loop.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @carmocca @justusschock @ananthsub @ninginthecloud @rohitgr7

### CODE REVIEW

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 1]
```diff
@@ -115,9 +116,10 @@ def advance(self, *args: Any, **kwargs: Any) -> None:
         )
         dl_max_batches = self._max_batches[dataloader_idx]
 
-        dl_outputs = self.epoch_loop.run(
-            dataloader, dataloader_idx if self.num_dataloaders > 1 else None, dl_max_batches
-        )
+        step_kwargs = OrderedDict()
+        if self.num_dataloaders > 1:
+            step_kwargs["dataloader_idx"] = dataloader_idx
+        dl_outputs = self.epoch_loop.run(dataloader, dl_max_batches, step_kwargs)
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:122]:
```suggestion
        kwargs = OrderedDict()
        if self.num_dataloaders > 1:
            kwargs["dataloader_idx"] = dataloader_idx
        dl_outputs = self.epoch_loop.run(dataloader, dl_max_batches, kwargs)
```

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 2]
```diff
@@ -269,23 +273,6 @@ def _on_evaluation_batch_end(self, output: Optional[STEP_OUTPUT], **kwargs: Any)
 
         self.trainer.logger_connector.on_batch_end()
 
-    def _build_kwargs(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int]) -> Dict[str, Union[Any, int]]:
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:272]: why don't we keep this? this was useful to override

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:272]: I am fine either way, but I liked this structure.

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:272]: Done

**Review by rohitgr7 — APPROVED** [review]
